library/src/main/java/com/getkeepsafe/relinker/ApkLibraryInstaller.java

```diff
@@ -0,0 +1,162 @@
+     * IO operations to ensure they succeed.
+     *
+     * @param context {@link Context} to describe the location of the installed APK file
+     * @param mappedLibraryName The mapped name of the library file to load
+     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    @Override
+    public void installLibrary(final Context context,
+                               final String[] abis,
+                               final String mappedLibraryName,
+                               final File destination,
+                               final ReLinkerInstance instance) {
+        ZipFile zipFile = null;
+        try {
+            final ApplicationInfo appInfo = context.getApplicationInfo();
+            int tries = 0;
+            while (tries++ < MAX_TRIES) {
+                try {
+                    zipFile = new ZipFile(new File(appInfo.sourceDir), ZipFile.OPEN_READ);
+                    break;
+                } catch (IOException ignored) {}
+            }
+
+            if (zipFile == null) {
+                instance.log("FATAL! Couldn't find application APK!");
+                return;
+            }
+
+            tries = 0;
+            while (tries++ < MAX_TRIES) {
+                String jniNameInApk = null;
+                ZipEntry libraryEntry = null;
+
+                for (final String abi : abis) {
+                    jniNameInApk = "lib" + File.separatorChar + abi + File.separatorChar
+                            + mappedLibraryName;
+                    libraryEntry = zipFile.getEntry(jniNameInApk);
+
+                    if (libraryEntry != null) {
+                        break;
+                    }
+                }
+
+                if (jniNameInApk != null) instance.log("Looking for %s in APK...", jniNameInApk);
+
+                if (libraryEntry == null) {
+                    // Does not exist in the APK
+                    if (jniNameInApk != null) {
+                        throw new MissingLibraryException(jniNameInApk);
+                    } else {
+                        throw new MissingLibraryException(mappedLibraryName);
+                    }
+                }
+
+                instance.log("Found %s! Extracting...", jniNameInApk);
+                try {
+                    if (!destination.exists() && !destination.createNewFile()) {
+                        continue;
+                    }
+                } catch (IOException ignored) {
+                    // Try again
+                    continue;
+                }
+
+                InputStream inputStream = null;
+                FileOutputStream fileOut = null;
+                try {
+                    inputStream = zipFile.getInputStream(libraryEntry);
+                    fileOut = new FileOutputStream(destination);
+                    copy(inputStream, fileOut);
+                } catch (FileNotFoundException e) {
+                    // Try again
```

Comment:
It doesn't die quietly. It will throw an `UnsatisfiedLinkError` in its parent function. If it reaches this part of the code, then we've deduced that the correct library file exists in the APK, however attempting to extract it has failed for transient reasons. I would just fail completely here and try again later (on app restart).

I've added a log message though, when that happens.